File: tagging_tools/LLVMMetadataTagger.py
import TaggingUtils
import logging
import sys
import yaml

from elftools.elf.constants import SH_FLAGS

def bytes_to_uint(it, size):
    data = []
    for _ in range(size):
        data.append(next(it))
    return int.from_bytes(data, byteorder='little', signed=False)

def round_up(x, align):
    return -((-x) & -align)

class LLVMMetadataTagger:
    PTR_SIZE = 4 # For instructions, not data

    metadata_ops = {
        "DMD_SET_BASE_ADDRESS_OP": 0x01,
        "DMD_TAG_ADDRESS_OP": 0x02,
        "DMD_TAG_ADDRESS_RANGE_OP": 0x03,
        "DMD_TAG_POLICY_SYMBOL": 0x04,
        "DMD_TAG_POLICY_RANGE": 0x05,
        "DMD_TAG_POLICY_SYMBOL_RANKED": 0x06,
        "DMD_TAG_POLICY_RANGE_RANKED": 0x07,
        "DMD_END_BLOCK": 0x08,
        "DMD_END_BLOCK_WEAK_DECL_HACK": 0x09,
        "DMD_FUNCTION_RANGE": 0xa
    }

    # LLVM feature mappings. See header file in LLVM
    tag_specifiers = {
        "DMT_CFI3L_VALID_TGT": 0x01,
        "DMT_STACK_PROLOGUE_AUTHORITY": 0x02,
        "DMT_STACK_EPILOGUE_AUTHORITY": 0x03,
        "DMT_FPTR_STORE_AUTHORITY": 0x04,
        "DMT_BRANCH_VALID_TGT": 0x05,
        "DMT_RET_VALID_TGT": 0x06,
        "DMT_RETURN_INSTR": 0x07,
        "DMT_CALL_INSTR": 0x08,
        "DMT_BRANCH_INSTR": 0x09,
        "DMT_FPTR_CREATE_AUTHORITY": 0x0a,
        "DMT_BEFORE_INLINE_ASM_INSNS": 0x0c,
        "DMT_AFTER_INLINE_ASM_INSNS": 0x0d
    }

    # Names policies will use to access features
    policy_map = {
        "call-tgt": {
            "tag_specifier": tag_specifiers["DMT_CFI3L_VALID_TGT"],
            "name": "llvm.CFI_Call-Tgt"
        },
        "branch-tgt": {
            "tag_specifier": tag_specifiers["DMT_BRANCH_VALID_TGT"],
            "name": "llvm.CFI_Branch-Tgt"
        },
        "return-tgt": {
            "tag_specifier": tag_specifiers["DMT_RET_VALID_TGT"],
            "name": "llvm.CFI_Return-Tgt"
        },
        "call-instr": {
            "tag_specifier": tag_specifiers["DMT_CALL_INSTR"],
            "name": "llvm.CFI_Call-Instr"
        },
        "branch-instr": {
            "tag_specifier": tag_specifiers["DMT_BRANCH_INSTR"],
            "name": "llvm.CFI_Branch-Instr"
        },
        "return-instr": {
            "tag_specifier": tag_specifiers["DMT_RETURN_INSTR"],
            "name": "llvm.CFI_Return-Instr"
        },
        "fptrcreate": {
            "tag_specifier": tag_specifiers["DMT_FPTR_CREATE_AUTHORITY"],
            "name": "llvm.CPI.FPtrCreate"
        },
        "fptrstore": {
            "tag_specifier": tag_specifiers["DMT_FPTR_STORE_AUTHORITY"],
            "name": "llvm.CPI.FPtrStore"
        },
        "prologue": {
            "tag_specifier": tag_specifiers["DMT_STACK_PROLOGUE_AUTHORITY"],
            "name": "llvm.Prologue"
        },
        "epilogue": {
            "tag_specifier": tag_specifiers["DMT_STACK_EPILOGUE_AUTHORITY"],
            "name": "llvm.Epilogue"
        },
        "before_inline_asm": {
            "tag_specifier": tag_specifiers["DMT_BEFORE_INLINE_ASM_INSNS"],
            "name": "llvm.Inline"
        },
        "after_inline_asm": {
            "tag_specifier": tag_specifiers["DMT_AFTER_INLINE_ASM_INSNS"],
            "name": "llvm.Inline"
        }
    }
    latest_inline_addr = -1
    latest_inline_addr_valid = False

    def __init__(self):
        self.needs_tag_cache = {}

    def policy_needs_tag(self, policy_inits, tag):

        if tag in self.needs_tag_cache:
            return self.needs_tag_cache[tag]

        d = policy_inits['Require']
        for item in tag.split("."):
            try:
                d = d[item]
            except KeyError:
                print("Don't need tag " + tag)
                self.needs_tag_cache[tag] = False
                return False

        self.needs_tag_cache[tag] = True
        print("Yes need tag " + tag)
        return True

    def add_code_section_ranges(self, elf_file, range_map):
        for s in elf_file.iter_sections():
            flags = s['sh_flags']
            if (flags & SH_FLAGS.SHF_ALLOC):
                start = s['sh_addr']
                end = start + s['sh_size']
                end = round_up(end, self.PTR_SIZE)
                if ((flags & (SH_FLAGS.SHF_ALLOC | SH_FLAGS.SHF_WRITE | SH_FLAGS.SHF_EXECINSTR)) ==
                    (SH_FLAGS.SHF_ALLOC | SH_FLAGS.SHF_EXECINSTR)):
                    range_map.add_range(start, end)

    def check_and_write_range(self, range_file, start, end, tag_specifier,
                            policy_inits, range_map):
        for policy, tags in self.policy_map.items():
            if self.policy_needs_tag(policy_inits, tags['name']) or self.policy_needs_tag(policy_inits, "llvm.Inline"):
                if tags['tag_specifier'] == tag_specifier:
#                     print(policy + ": start " + hex(start) + ", end " + hex(end));
                    if tags['name'] == "llvm.Prologue":
                        self.latest_inline_addr_valid = False
                        self.latest_inline_addr = start + 4;
                    if tags['name'] == "llvm.Epilogue" and self.latest_inline_addr > 0 and self.latest_inline_addr_valid:
                            print("Should tag between " + hex(self.latest_inline_addr) + " and epilogue at " + hex(start) + " as INLINE_ASM")
                            range_file.write_range(self.latest_inline_addr, start, "llvm.Inline")
                            range_map.add_range(self.latest_inline_addr, start, "llvm.Inline")
                            self.latest_inline_addr = -1
                            self.latest_inline_addr_valid = False

                    if tags['name'] == "llvm.Inline":
                        if tag_specifier == self.tag_specifiers["DMT_BEFORE_INLINE_ASM_INSNS"]:
                            self.latest_inline_addr_valid = True
                            self.latest_inline_addr = start + 4
                        elif self.latest_inline_addr > 0 and tag_specifier == self.tag_specifiers["DMT_AFTER_INLINE_ASM_INSNS"]:
                            # Do something
                            if self.latest_inline_addr_valid:
                                print("Should tag between " + hex(self.latest_inline_addr) + " and " + hex(start) + " as INLINE_ASM")
                            else:
                                print("Should tag between prologue at " + hex(self.latest_inline_addr) + " and " + hex(start) + " as INLINE_ASM")
                            range_file.write_range(self.latest_inline_addr, start, tags['name'])
                            range_map.add_range(self.latest_inline_addr, start, tags['name'])
                            self.latest_inline_addr = -1
                            self.latest_inline_addr_valid = False
                            pass
                        elif tag_specifier == self.tag_specifiers["DMT_BEFORE_INLINE_ASM_INSNS"]:
                            print("Invalid double before tag " + hex(self.latest_inline_addr) + " and " + hex(start) + ".")
                            # Maybe if we hit prologue or epilogue, we should reset vakyes, too.
                        else:
                            print("Invalid after inline tag with no before " + hex(start) + ".")
                    elif self.policy_needs_tag(policy_inits, tags['name']):
                        range_file.write_range(start, end, tags['name'])
                        range_map.add_range(start, end, tags['name'])

    def generate_policy_ranges(self, elf_file, range_file, policy_inits):
        metadata = elf_file.get_section_by_name(b'.dover_metadata')
        if not metadata:
            metadata = elf_file.get_section_by_name(".dover_metadata")
        assert metadata, "No metadata found in ELF file!"
        metadata = metadata.data()
        assert metadata[0] == self.metadata_ops['DMD_SET_BASE_ADDRESS_OP'], "Invalid metadata found in ELF file!"

        it = iter(metadata)

        range_map = TaggingUtils.RangeMap()

        for byte in it:
            if (byte == self.metadata_ops['DMD_SET_BASE_ADDRESS_OP']):
                base_address = bytes_to_uint(it, 8) #apparently GCC emits a 64-bit base
                logging.debug("new base address is " + hex(base_address) + "\n")
            elif (byte == self.metadata_ops['DMD_TAG_ADDRESS_OP']):
                address = bytes_to_uint(it, self.PTR_SIZE) + base_address
                tag_specifier = bytes_to_uint(it, 1)
                logging.debug("tag is " + hex(tag_specifier) +
                            " at address " + hex(address) + '\n')

                self.check_and_write_range(range_file, address, address + self.PTR_SIZE,
                                    tag_specifier, policy_inits, range_map)

            elif (byte == self.metadata_ops['DMD_TAG_ADDRESS_RANGE_OP']):
                start_address = bytes_to_uint(it, self.PTR_SIZE) + base_address
                end_address = bytes_to_uint(it, self.PTR_SIZE) + base_address
                tag_specifier = bytes_to_uint(it, 1)
                logging.debug("tag is " + hex(tag_specifier) +
                            " for address range " +
                            hex(start_address) + ":" + hex(end_address) + '\n')

                self.check_and_write_range(range_file, start_address, end_address,
                                    tag_specifier, policy_inits, range_map)

            elif (byte == self.metadata_ops['DMD_TAG_POLICY_SYMBOL']):
                logging.critical("Saw policy symbol!\n")
                sys.exit(-1)
            elif (byte == self.metadata_ops['DMD_TAG_POLICY_RANGE']):
                logging.critical("Saw policy range!\n")
                sys.exit(-1)
                for _ in range(self.PTR_SIZE*3):
                    #skip start, end, & 32-bit tag-type
                    next(it)
            elif (byte == self.metadata_ops['DMD_TAG_POLICY_SYMBOL_RANKED']):
                logging.critical("Saw policy symbol ranked\n")
                sys.exit(-1)
            elif (byte == self.metadata_ops['DMD_TAG_POLICY_RANGE_RANKED']):
                sys.exit(-1)
                logging.critical("Saw policy symbol range ranked\n")
                #skip start, end ,tag category, rank, tag type
                for _ in range(self.PTR_SIZE*5):
                    next(it)
            elif (byte == self.metadata_ops['DMD_END_BLOCK_WEAK_DECL_HACK']):
                logging.critical("saw end weak decl tag!\n")
                sys.exit(-1)
            elif (byte == self.metadata_ops['DMD_END_BLOCK']):
                end_address = bytes_to_uint(it, self.PTR_SIZE)
                logging.debug("saw end block tag range = " + hex(base_address) +
                            ":" + hex(base_address + end_address))
                range_map.add_range(base_address, base_address + end_address, "COMPILER_GENERATED")
            elif (byte == self.metadata_ops['DMD_FUNCTION_RANGE']):
                start_address = bytes_to_uint(it, self.PTR_SIZE) + base_address
                end_address = bytes_to_uint(it, self.PTR_SIZE) + base_address
                logging.debug("saw function range = " + hex(start_address) +
                            ":" + hex(end_address))
                range_map.add_range(start_address, end_address, "COMPILER_GENERATED")
            else:
                logging.debug("Error: found unknown byte in metadata!" + hex(byte) + "\n")
                sys.exit(-1)

        # tag NoCFI for anything not specifically noted by llvm
        if 'NoCFI' in policy_inits['Require']['llvm']:

            code_range_map = TaggingUtils.RangeMap()
            self.add_code_section_ranges(elf_file, code_range_map)

            for (start, end, tags) in code_range_map:
                for s in range(start, end, self.PTR_SIZE):
                    e = s + self.PTR_SIZE
                    if (s, e, tags) not in range_map:
                        range_file.write_range(s, e, "llvm.NoCFI")

        return range_map

File: tagging_tools/llvm_metadata_tagger.cc
#include <algorithm>
#include <cstdint>
#include <gelf.h>
#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <unistd.h>
#include <yaml-cpp/yaml.h>
#include "elf_loader.h"
#include "llvm_metadata_tagger.h"
#include "range_map.h"

namespace policy_engine {

uint64_t round_up(uint64_t x, uint64_t align) {
  return ((x + align - 1)/align)*align;
}

const std::map<std::string, uint8_t> llvm_metadata_tagger_t::metadata_ops = {
  {"DMD_SET_BASE_ADDRESS_OP", 0x1},
  {"DMD_TAG_ADDRESS_OP", 0x2},
  {"DMD_TAG_ADDRESS_RANGE_OP", 0x3},
  {"DMD_TAG_POLICY_SYMBOL", 0x4},
  {"DMD_TAG_POLICY_RANGE", 0x5},
  {"DMD_TAG_POLICY_SYMBOL_RANKED", 0x6},
  {"DMD_TAG_POLICY_RANGE_RANKED", 0x7},
  {"DMD_END_BLOCK", 0x8},
  {"DMD_END_BLOCK_WEAK_DECL_HACK", 0x9},
  {"DMD_FUNCTION_RANGE", 0xa}
};

const std::map<std::string, uint8_t> llvm_metadata_tagger_t::tag_specifiers = {
  {"DMT_CFI3L_VALID_TGT", 0x1},
  {"DMT_STACK_PROLOGUE_AUTHORITY", 0x2},
  {"DMT_STACK_EPILOGUE_AUTHORITY", 0x3},
  {"DMT_FPTR_STORE_AUTHORITY", 0x4},
  {"DMT_BRANCH_VALID_TGT", 0x5},
  {"DMT_RET_VALID_TGT", 0x6},
  {"DMT_RETURN_INSTR", 0x7},
  {"DMT_CALL_INSTR", 0x8},
  {"DMT_BRANCH_INSTR", 0x9},
  {"DMT_FPTR_CREATE_AUTHORITY", 0xa}
};

const std::map<std::string, std::map<std::string, std::string>> llvm_metadata_tagger_t::policy_map = {
  {"call-tgt", {
    {"tag_specifier", "DMT_CFI3L_VALID_TGT"},
    {"name", "llvm.CFI_Call-Tgt"}
  }},
  {"branch-tgt", {
    {"tag_specifier", "DMT_BRANCH_VALID_TGT"},
    {"name", "llvm.CFI_Branch-Tgt"}
  }},
  {"return-tgt", {
    {"tag_specifier", "DMT_RET_VALID_TGT"},
    {"name", "llvm.CFI_Return-Tgt"}
  }},
  {"call-instr", {
    {"tag_specifier", "DMT_CALL_INSTR"},
    {"name", "llvm.CFI_Call-Instr"}
  }},
  {"branch-instr", {
    {"tag_specifier", "DMT_BRANCH_INSTR"},
    {"name", "llvm.CFI_Branch-Instr"}
  }},
  {"return-instr", {
    {"tag_specifier", "DMT_RETURN_INSTR"},
    {"name", "llvm.CFI_Return-Instr"}
  }},
  {"fptrcreate", {
    {"tag_specifier", "DMT_FPTR_CREATE_AUTHORITY"},
    {"name", "llvm.CPI.FPtrCreate"}
  }},
  {"fptrstore", {
    {"tag_specifier", "DMT_FPTR_STORE_AUTHORITY"},
    {"name", "llvm.CPI.FPtrStore"}
  }},
  {"prologue", {
    {"tag_specifier", "DMT_STACK_PROLOGUE_AUTHORITY"},
    {"name", "llvm.Prologue"}
  }},
  {"epilogue", {
    {"tag_specifier", "DMT_STACK_EPILOGUE_AUTHORITY"},
    {"name", "llvm.Epilogue"}
  }}
};

bool llvm_metadata_tagger_t::policy_needs_tag(const YAML::Node& node, const std::string& tag) {
  if (needs_tag_cache.find(tag) != needs_tag_cache.end())
    return needs_tag_cache[tag];
  
  std::string::size_type pos = tag.find('.');
  std::string item = tag.substr(0, pos);
  if (!node[item]) {
    needs_tag_cache[tag] = false;
    return false;
  } else if (pos == std::string::npos) {
    needs_tag_cache[tag] = true;
    return true;
  } else {
    bool needs_tag = policy_needs_tag(node[item], tag.substr(pos + 1, std::string::npos));
    needs_tag_cache[tag] = needs_tag;
    return needs_tag;
  }
}

void llvm_metadata_tagger_t::add_code_section_ranges(const elf_image_t& ef, range_map_t& range_map) {
  for (int i = 0; i < ef.sections.size(); i++) {
    if (ef.sections[i].flags & SHF_ALLOC) {
      uint64_t start = ef.sections[i].address;
      uint64_t end = round_up(start + ef.sections[i].size, PTR_SIZE);
      if ((ef.sections[i].flags & (SHF_ALLOC | SHF_WRITE | SHF_EXECINSTR)) == (SHF_ALLOC | SHF_EXECINSTR)) {
        err.info("saw code range = %#lx:%#lx\n", start, end);
        range_map.add_range(start, end);
      }
    }
  }
}

void llvm_metadata_tagger_t::check_and_add_range(range_map_t& range_map, uint64_t start, uint64_t end, uint8_t tag_specifier, const YAML::Node& policy_inits) {
  for (const auto& [ policy, tags ] : policy_map) {
    if (policy_needs_tag(policy_inits["Require"], tags.at("name"))) {
      if (tag_specifiers.at(tags.at("tag_specifier")) == tag_specifier) {
        err.info("saw tag %s = %#lx:%#lx\n", tags.at("name").c_str(), start, end);
        range_map.add_range(start, end, tags.at("name"));
      }
    }
  }
}

void llvm_metadata_tagger_t::add_policy_ranges(range_map_t& range_map, const elf_image_t& ef, const YAML::Node& policy_inits) {
  auto metadata_section = std::find_if(ef.sections.begin(), ef.sections.end(), [](const elf_section_t& s){ return s.name == ".dover_metadata"; });
  if (metadata_section == ef.sections.end())
    throw std::runtime_error("no metadata found in ELF file");
  uint8_t* metadata = reinterpret_cast<uint8_t*>(metadata_section->data);
  if (metadata[0] != metadata_ops.at("DMD_SET_BASE_ADDRESS_OP"))
    throw std::runtime_error("invalid metadata found in ELF file");
  
  uint64_t base_address = 0;
  for (int i = 0; i < metadata_section->size;) {
    // Don't increment i at the end of the loop because it should point to the next op after the if block
    uint8_t op = metadata[i++];
    if (op == metadata_ops.at("DMD_SET_BASE_ADDRESS_OP")) {
      base_address = 0;
      for (int j = 0; j < 8; j++, i++)
        base_address += metadata[i] << (j*8);
      err.info("new base address is %#x\n", base_address);
    } else if (op == metadata_ops.at("DMD_TAG_ADDRESS_OP")) {
      uint64_t address = base_address;
      for (int j = 0; j < PTR_SIZE; j++, i++)
        address += metadata[i] << (j*8);
      uint8_t tag_specifier = metadata[i++];
      err.info("tag is %#x at address %#lx\n", tag_specifier, address);
      check_and_add_range(range_map, address, address + PTR_SIZE, tag_specifier, policy_inits);
    } else if (op == metadata_ops.at("DMD_TAG_ADDRESS_OP")) {
      uint64_t start_address = base_address, end_address = base_address;
      for (int j = 0; j < PTR_SIZE; j++, i++)
        start_address += metadata[i] << (j*8);
      for (int j = 0; j < PTR_SIZE; j++, i++)
        end_address += metadata[i] << (j*8);
      uint8_t tag_specifier = metadata[i++];
      err.info("tag is %#x for address range %#lx:%#lx\n", tag_specifier, start_address, end_address);
      check_and_add_range(range_map, start_address, end_address, tag_specifier, policy_inits);
    } else if (op == metadata_ops.at("DMD_END_BLOCK")) {
      uint64_t end_address = 0;
      for (int j = 0; j < PTR_SIZE; j++, i++)
        end_address += metadata[i] << (j*8);
      err.info("saw end block tag range = %#lx:%#lx\n", base_address, base_address + end_address);
    } else if (op == metadata_ops.at("DMD_FUNCTION_RANGE")) {
      uint64_t start_address = base_address, end_address = base_address;
      for (int j = 0; j < PTR_SIZE; j++, i++)
        start_address += metadata[i] << (j*8);
      for (int j = 0; j < PTR_SIZE; j++, i++)
        end_address += metadata[i] << (j*8);
      err.info("saw function range = %#lx:%#lx\n", start_address, end_address);
    } else if (op == metadata_ops.at("DMD_TAG_POLICY_SYMBOL")) {
      throw std::runtime_error("saw policy symbol");
    } else if (op == metadata_ops.at("DMD_TAG_POLICY_RANGE")) {
      throw std::runtime_error("saw policy range");
    } else if (op == metadata_ops.at("DMD_TAG_POLICY_SYMBOL_RANKED")) {
      throw std::runtime_error("saw policy symbol ranked");
    } else if (op == metadata_ops.at("DMD_TAG_POLICY_RANGE_RANKED")) {
      throw std::runtime_error("saw policy range ranked");
    } else if (op == metadata_ops.at("DMD_END_BLOCK_WEAK_DECL_HACK")) {
      throw std::runtime_error("saw end weak decl tag");
    } else {
      throw std::runtime_error("found unknown byte in metadata: " + std::to_string(metadata[i]));
    }
  }

  if (policy_inits["Require"]["llvm"]["NoCFI"]) {
    range_map_t code_range_map;
    add_code_section_ranges(ef, code_range_map);
    for (auto& [ range, tags ] : code_range_map) {
      for (uint64_t s = range.start; s < range.end; s += PTR_SIZE) {
        uint64_t e = s + PTR_SIZE;
        if (!range_map.contains(tagged_range_t{{s, e}, tags})) {
          err.info("llvm.NoCFI range = %lx:%lx\n", s, e);
          range_map.add_range(s, e, "llvm.NoCFI");
        }
      }
    }
  }
}

} // namespace policy_engine